### PR TITLE
Add file-split-ftp Sample

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/INTSAMPLES-149

### CODE REVIEW

**Comment by artembilan** [comment]
Well, welcome back and your test have been failed :smile: 
```
org.springframework.integration.samples.filesplit.ApplicationTests > testFailure FAILED

    java.lang.IllegalStateException

        Caused by: org.springframework.beans.factory.BeanCreationException

            Caused by: org.springframework.beans.factory.BeanCreationException

                Caused by: org.springframework.beans.factory.NoSuchBeanDefinitionException

org.springframework.integration.samples.filesplit.ApplicationTests > testSuccess FAILED

    java.lang.IllegalStateException

        Caused by: org.springframework.beans.factory.BeanCreationException

            Caused by: org.springframework.beans.factory.BeanCreationException

                Caused by: org.springframework.beans.factory.NoSuchBeanDefinitionException
```

**Comment by garyrussell** [comment]
Strange, runs fine with gradle for me locally; re-pushed with top level README update. Let's see...

File: applications/file-split-ftp/src/main/java/org/springframework/integration/samples/filesplit/Application.java [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+	}
+
+	@Autowired
+	private MailProperties mailProperties;
+
+	@Bean
+	public IntegrationFlow fromFile() {
+		return IntegrationFlows.from(Files.inboundAdapter(new File("/tmp/in"))
+					.patternFilter("*.txt"), e -> e.poller(Pollers.fixedDelay(5000)))
+				.enrichHeaders(h -> h.header(MessageHeaders.ERROR_CHANNEL, "tfrErrors.input"))
```
Comment (artembilan) [inline applications/file-split-ftp/src/main/java/org/springframework/integration/samples/filesplit/Application.java]: https://jira.spring.io/browse/INT-4113.
I mean add comment here.
Although I think after `getComponentsToRegister()` for the `PollerSpec` we can provide some out-of-the-box workaround just with `errorChannel(MessageChannelString)`
But right: that is a different story...

File: applications/file-split-ftp/src/main/java/org/springframework/integration/samples/filesplit/Application.java [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+						.subscribe(sf -> sf.<FileSplitter.FileMarker, File>transform(p -> new File("/tmp/out/009.txt"))
+								.enrichHeaders(h -> h.header(FileHeaders.FILENAME, "009.txt", true))
+								.handleWithAdapter(a -> a.ftp(ftp3()).remoteDirectory("foo"), e -> e.id("ftp009")))
+
+						// send an email
+						.subscribe(sf -> sf.<FileSplitter.FileMarker, String>transform(p -> p.getFilePath())
+								.enrichHeaders(Mail.headers()
+										.subject("File successfully split and transferred")
+										.from("foo@bar")
+										.toFunction(m -> new String[] {"bar@baz"}))
```
Comment (artembilan) [inline applications/file-split-ftp/src/main/java/org/springframework/integration/samples/filesplit/Application.java]: Yep! These are up to `@ConfigurationProperties`. Our own custom one.
I mean for the real mail server tests.
But I see we have so many of options to expose...

Maybe we can live without that for now...

File: applications/file-split-ftp/src/main/java/org/springframework/integration/samples/filesplit/Application.java (hunk 1 above)
Comment (garyrussell) [inline applications/file-split-ftp/src/main/java/org/springframework/integration/samples/filesplit/Application.java]: I don't want to pollute the code here with a TODO: will add a comment to the JIRA.

File: applications/file-split-ftp/src/main/java/org/springframework/integration/samples/filesplit/Application.java [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+	}
+
+	/**
+	 * Process lines; append (no flush) to the appropriate file.
+	 * @return the flow.
+	 */
+	@Bean
+	public IntegrationFlow lines() {
+		return f -> f.enrichHeaders(h -> h.headerExpression(
+						FileHeaders.FILENAME, "payload.substring(1, 4) + '.txt'", true))
```
Comment (artembilan) [inline applications/file-split-ftp/src/main/java/org/springframework/integration/samples/filesplit/Application.java]: WDYT, if we change this `.enrichHeaders()` just to the `.fileNameExpression("payload.substring(1, 4) + '.txt'")`?

Although it doesn't work with your test somehow... But I see in debug the proper expression and expected file name result.

File: applications/file-split-ftp/src/main/java/org/springframework/integration/samples/filesplit/Application.java [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+	 */
+	@Bean
+	public IntegrationFlow markers() {
+		return f -> f.<FileSplitter.FileMarker>filter(m -> m.getMark().equals(FileSplitter.FileMarker.Mark.END),
+						e -> e.id("markerFilter"))
+				.publishSubscribeChannel(s -> s
+
+						// first trigger file flushes
+						.subscribe(sf -> sf.transform("'/tmp/out/.*\\.txt'", e -> e.id("toTriggerPattern"))
+										.handle("fileOut.handler", "trigger", e -> e.id("flusher")))
```
Comment (artembilan) [inline applications/file-split-ftp/src/main/java/org/springframework/integration/samples/filesplit/Application.java]: Having the problem with bean definitions order on Linux we should move `Files.outboundAdapter()` to the separate `@Bean`. At least until the fix for https://github.com/spring-projects/spring-integration-java-dsl/issues/119

File: applications/file-split-ftp/src/main/java/org/springframework/integration/samples/filesplit/Application.java (hunk 3 above)
Comment (artembilan) [inline applications/file-split-ftp/src/main/java/org/springframework/integration/samples/filesplit/Application.java]: Gary, any ideas why it doesn't work with just `.fileNameExpression()` and I'll merge with the fix for handler bean access.
Thanks

Comment (garyrussell) [inline applications/file-split-ftp/src/main/java/org/springframework/integration/samples/filesplit/Application.java]: I will look.

Comment (garyrussell) [inline applications/file-split-ftp/src/main/java/org/springframework/integration/samples/filesplit/Application.java]: It worked fine for me with the file name expression. I've built on linux too. Let's see what our friend Travis thinks.

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-integration-samples/commit/2c1a070930b54b724d6377f85211e872bf2c6a24.

See my polishing to the tests according Windows compatibility.

**Pull request closed without merging**